File: vless_app/singbox_config.py
from typing import Any, Dict, List
import sys

from .models import Server

LOCAL_PORT = 7890


def _mixed_inbound(port: int = LOCAL_PORT) -> Dict[str, Any]:
    return {
        "type": "mixed",
        "tag": "mixed-in",
        "listen": "127.0.0.1",
        "listen_port": port,
    }


def _tun_inbound() -> Dict[str, Any]:
    return {
        "type": "tun",
        "tag": "tun-in",
        "interface_name": "grey-tun" if sys.platform == "win32" else "tun0",
        "address": ["172.19.0.1/30"],
        "auto_route": True,
        "strict_route": True,
        "stack": "system",
        "sniff": True,
        "sniff_override_destination": True,
    }


def _dns_block() -> Dict[str, Any]:
    return {
        "servers": [
            {"tag": "remote", "address": "tls://8.8.8.8", "detour": "proxy"},
            {"tag": "local", "address": "223.5.5.5", "detour": "direct"},
        ],
        "rules": [
            {"outbound": "any", "server": "remote"},
        ],
        "final": "remote",
        "strategy": "prefer_ipv4",
    }


def _route_block() -> Dict[str, Any]:
    return {
        "auto_detect_interface": True,
        "rules": [
            {"protocol": "dns", "action": "hijack-dns"},
        ],
        "final": "proxy",
    }


def _transport(server: Server) -> Dict[str, Any] | None:
    network = server.params.get("network") or "tcp"
    if network in ("tcp", "none", ""):
        return None

    transport: Dict[str, Any] = {"type": network}

    if network == "ws":
        transport["path"] = server.params.get("path") or "/"
        host = server.params.get("host") or server.params.get("host_header")
        if host:
            transport["headers"] = {"Host": host}
    elif network == "grpc":
        transport["service_name"] = server.params.get("serviceName") or server.params.get("service_name") or ""
    elif network == "http":
        transport["path"] = server.params.get("path") or "/"
        host = server.params.get("host") or server.params.get("host_header")
        if host:
            transport["host"] = [host]

    return transport


def _tls(server: Server) -> Dict[str, Any] | None:
    security = (server.params.get("security") or server.params.get("tls") or "").lower()
    if security not in ("tls", "reality", "xtls"):
        return None

    tls: Dict[str, Any] = {
        "enabled": True,
        "server_name": server.params.get("sni") or server.host,
    }

    fp = server.params.get("fp")
    if fp:
        tls["utls"] = {"enabled": True, "fingerprint": fp}

    if security == "reality":
        tls["reality"] = {
            "enabled": True,
            "public_key": server.params.get("pbk") or "",
            "short_id": server.params.get("sid") or "",
        }

    return tls


def _vless_outbound(server: Server) -> Dict[str, Any]:
    outbound: Dict[str, Any] = {
        "type": "vless",
        "tag": "proxy",
        "server": server.host,
        "server_port": server.port,
        "uuid": server.params.get("uuid") or "",
    }

    flow = server.params.get("flow")
    if flow:
        outbound["flow"] = flow

    transport = _transport(server)
    if transport:
        outbound["transport"] = transport

    tls = _tls(server)
    if tls:
        outbound["tls"] = tls

    return outbound


def _vmess_outbound(server: Server) -> Dict[str, Any]:
    outbound: Dict[str, Any] = {
        "type": "vmess",
        "tag": "proxy",
        "server": server.host,
        "server_port": server.port,
        "uuid": server.params.get("uuid") or "",
        "security": server.params.get("security") or "auto",
        "alter_id": int(server.params.get("alter_id") or 0),
    }

    transport = _transport(server)
    if transport:
        outbound["transport"] = transport

    tls = _tls(server)
    if tls:
        outbound["tls"] = tls

    return outbound


def _trojan_outbound(server: Server) -> Dict[str, Any]:
    outbound: Dict[str, Any] = {
        "type": "trojan",
        "tag": "proxy",
        "server": server.host,
        "server_port": server.port,
        "password": server.params.get("password") or "",
    }

    transport = _transport(server)
    if transport:
        outbound["transport"] = transport

    tls = _tls(server) or {"enabled": True, "server_name": server.params.get("sni") or server.host}
    outbound["tls"] = tls
    return outbound


def _shadowsocks_outbound(server: Server) -> Dict[str, Any]:
    return {
        "type": "shadowsocks",
        "tag": "proxy",
        "server": server.host,
        "server_port": server.port,
        "method": server.params.get("method") or "aes-256-gcm",
        "password": server.params.get("password") or "",
    }


def build_config(server: Server, local_port: int = LOCAL_PORT, tun_mode: bool = False) -> Dict[str, Any]:
    builders = {
        "vless": _vless_outbound,
        "vmess": _vmess_outbound,
        "trojan": _trojan_outbound,
        "shadowsocks": _shadowsocks_outbound,
    }

    builder = builders.get(server.protocol)
    if not builder:
        raise ValueError(f"Протокол {server.protocol} пока не поддерживается")

    inbounds: List[Dict[str, Any]] = []
    if tun_mode:
        inbounds.append(_tun_inbound())
    # Локальный HTTP/SOCKS — всегда (curl -x, браузер, системный прокси)
    inbounds.append(_mixed_inbound(local_port))

    return {
        "log": {"level": "warn"},
        "dns": _dns_block(),
        "inbounds": inbounds,
        "outbounds": [
            builder(server),
            {"type": "direct", "tag": "direct"},
        ],
        "route": _route_block(),
    }

File: grey_vless/lib/services/singbox_config.dart
import 'dart:io';

import '../models/server.dart';
import '../models/tunnel_mode.dart';
import '../platform/installed_apps.dart';

class SingboxConfigBuilder {
  static const localPort = 7890;

  static Map<String, dynamic> _mixedInbound() => {
        'type': 'mixed',
        'tag': 'mixed-in',
        'listen': '127.0.0.1',
        'listen_port': localPort,
        'sniff': true,
        'sniff_override_destination': true,
        'set_system_proxy': false,
      };

  static Map<String, dynamic> _tunInbound() {
    if (Platform.isAndroid) {
      return {
        'type': 'tun',
        'tag': 'tun-in',
        'address': ['172.19.0.1/30'],
        'auto_route': true,
        'strict_route': true,
        'stack': 'gvisor',
        'sniff': true,
        'sniff_override_destination': true,
      };
    }
    // Windows: WinTun; имя tun0 часто конфликтует — используем grey-tun.
    final iface = Platform.isWindows ? 'grey-tun' : 'tun0';
    return {
      'type': 'tun',
      'tag': 'tun-in',
      'interface_name': iface,
      'address': ['172.19.0.1/30'],
      'auto_route': true,
      'strict_route': true,
      'stack': 'system',
      'sniff': true,
      'sniff_override_destination': true,
    };
  }

  static Map<String, dynamic> build(
    VpnServer server, {
    bool tunMode = false,
    TunnelMode tunnelMode = TunnelMode.fullVpn,
    List<String> tunnelAppIds = const [],
  }) {
    final inbounds = <Map<String, dynamic>>[];
    // Per-app на desktop требует TUN, чтобы видеть процесс источника.
    final useTun = (tunMode || tunnelMode.needsAppList) && !Platform.isAndroid;
    if (useTun) {
      inbounds.add(_tunInbound());
    }
    inbounds.add(_mixedInbound());

    return {
      'log': {'level': Platform.isAndroid ? 'info' : 'warn'},
      'dns': _dnsBlock(),
      'inbounds': inbounds,
      'outbounds': [
        _outbound(server),
        {'type': 'direct', 'tag': 'direct'},
      ],
      'route': _routeBlock(tunnelMode: tunnelMode, tunnelAppIds: tunnelAppIds),
    };
  }

  static Map<String, dynamic> _routeBlock({
    required TunnelMode tunnelMode,
    required List<String> tunnelAppIds,
  }) {
    final rules = <Map<String, dynamic>>[
      {'protocol': 'dns', 'action': 'hijack-dns'},
    ];

    // Свой клиент и ядро — всегда direct (анти-петля).
    if (!Platform.isAndroid) {
      rules.add({
        'process_name': ['grey_vless.exe', 'grey_vless', 'sing-box.exe', 'sing-box'],
        'outbound': 'direct',
      });
    }

    if (Platform.isAndroid) {
      rules.add({
        'domain_suffix': [
          'telegram.org',
          't.me',
          'telegra.ph',
          'telegram.me',
          'tdesktop.com',
          'telesco.pe',
        ],
        'outbound': 'proxy',
      });
      // Per-app на Android делает VpnService; в конфиге final=proxy.
      return {
        'rules': rules,
        'final': 'proxy',
      };
    }

    // Desktop per-app через process_name / process_path (нужен find_process).
    if (tunnelMode.needsAppList && tunnelAppIds.isNotEmpty) {
      final matchers = InstalledApps.matchersFromIds(tunnelAppIds);
      final processRule = <String, dynamic>{};
      if (matchers.names.isNotEmpty) {
        processRule['process_name'] = matchers.names;
      }
      if (matchers.paths.isNotEmpty) {
        processRule['process_path'] = matchers.paths;
      }
      if (processRule.isNotEmpty) {
        if (tunnelMode == TunnelMode.selectedApps) {
          // Только выбранные → proxy, остальное напрямую.
          processRule['outbound'] = 'proxy';
          rules.add(processRule);
          return {
            'auto_detect_interface': true,
            'find_process': true,
            'rules': rules,
            'final': 'direct',
          };
        }
        if (tunnelMode == TunnelMode.bypassApps) {
          // Выбранные → direct, остальное → proxy.
          processRule['outbound'] = 'direct';
          rules.add(processRule);
          return {
            'auto_detect_interface': true,
            'find_process': true,
            'rules': rules,
            'final': 'proxy',
          };
        }
      }
    }

    return {
      'auto_detect_interface': true,
      'find_process': tunnelMode.needsAppList,
      'rules': rules,
      'final': 'proxy',
    };
  }

  static Map<String, dynamic> _dnsBlock() {
    // DNS через direct — иначе sing-box не может резолвить адрес прокси до поднятия туннеля.
    return {
      'servers': [
        {'tag': 'google', 'address': '8.8.8.8', 'detour': 'direct'},
        {'tag': 'cloudflare', 'address': '1.1.1.1', 'detour': 'direct'},
      ],
      'final': 'google',
      'strategy': 'prefer_ipv4',
    };
  }

  static Map<String, dynamic> _outbound(VpnServer server) {
    switch (server.protocol) {
      case 'vless':
        return _vless(server);
      case 'vmess':
        return _vmess(server);
      case 'trojan':
        return _trojan(server);
      case 'shadowsocks':
        return _shadowsocks(server);
      default:
        throw UnsupportedError('Протокол ${server.protocol} не поддерживается');
    }
  }

  static Map<String, dynamic>? _transport(VpnServer server) {
    final network = server.params['network'] ?? 'tcp';
    if (network == 'tcp' || network == 'none' || network.isEmpty) return null;
    final transport = <String, dynamic>{'type': network};
    if (network == 'ws') {
      transport['path'] = server.params['path']?.isNotEmpty == true ? server.params['path'] : '/';
      final host = server.params['host'] ?? server.params['host_header'];
      if (host != null && host.isNotEmpty) {
        transport['headers'] = {'Host': host};
      }
    } else if (network == 'grpc') {
      transport['service_name'] = server.params['serviceName'] ?? '';
    }
    return transport;
  }

  static Map<String, dynamic>? _tls(VpnServer server) {
    final security = (server.params['security'] ?? server.params['tls'] ?? '').toLowerCase();
    if (!['tls', 'reality', 'xtls'].contains(security)) return null;
    final tls = <String, dynamic>{
      'enabled': true,
      'server_name': (server.params['sni']?.isNotEmpty == true) ? server.params['sni'] : server.host,
    };
    final fp = server.params['fp'];
    if (fp != null && fp.isNotEmpty) {
      tls['utls'] = {'enabled': true, 'fingerprint': fp};
    }
    if (security == 'reality') {
      tls['reality'] = {
        'enabled': true,
        'public_key': server.params['pbk'] ?? '',
        'short_id': server.params['sid'] ?? '',
      };
    }
    return tls;
  }

  static Map<String, dynamic> _vless(VpnServer server) {
    final outbound = <String, dynamic>{
      'type': 'vless',
      'tag': 'proxy',
      'server': server.host,
      'server_port': server.port,
      'uuid': server.params['uuid'] ?? '',
    };
    final flow = server.params['flow'];
    if (flow != null && flow.isNotEmpty) outbound['flow'] = flow;
    final transport = _transport(server);
    if (transport != null) outbound['transport'] = transport;
    final tls = _tls(server);
    if (tls != null) outbound['tls'] = tls;
    return outbound;
  }

  static Map<String, dynamic> _vmess(VpnServer server) {
    final outbound = <String, dynamic>{
      'type': 'vmess',
      'tag': 'proxy',
      'server': server.host,
      'server_port': server.port,
      'uuid': server.params['uuid'] ?? '',
      'security': server.params['security'] ?? 'auto',
      'alter_id': int.tryParse(server.params['alter_id'] ?? '0') ?? 0,
    };
    final transport = _transport(server);
    if (transport != null) outbound['transport'] = transport;
    final tls = _tls(server);
    if (tls != null) outbound['tls'] = tls;
    return outbound;
  }

  static Map<String, dynamic> _trojan(VpnServer server) {
    final outbound = <String, dynamic>{
      'type': 'trojan',
      'tag': 'proxy',
      'server': server.host,
      'server_port': server.port,
      'password': server.params['password'] ?? '',
    };
    final transport = _transport(server);
    if (transport != null) outbound['transport'] = transport;
    outbound['tls'] = _tls(server) ??
        {
          'enabled': true,
          'server_name': (server.params['sni']?.isNotEmpty == true) ? server.params['sni'] : server.host,
        };
    return outbound;
  }

  static Map<String, dynamic> _shadowsocks(VpnServer server) {
    return {
      'type': 'shadowsocks',
      'tag': 'proxy',
      'server': server.host,
      'server_port': server.port,
      'method': server.params['method'] ?? 'aes-256-gcm',
      'password': server.params['password'] ?? '',
    };
  }
}
